File: php/matching-brackets/MatchingBrackets.php
<?php
// Matching Brackets Exercise.

declare(strict_types=1);


// Check if the given text has matching brackets. The count and ordering should match.
// Only checks (, [, and {
// @param $input: The text to check.
// @returns: True if the braces match and false otherwise.
function brackets_match(string $input): bool
{
    $start = "[({";
    $finish = "])}";
    $nesting = array();
    if (strlen($input) > 0) {
        foreach(str_split($input) as $char) {
            if (str_contains($start, $char)) {
                $nesting[] = $char;
            } else if (str_contains($finish, $char)) {
                if (count($nesting) == 0) {
                    return false;
                }                
                $match = end($nesting);
                if (strpos($start, $match) == strpos($finish, $char)) {
                    array_pop($nesting);
                } else {
                    return false;
                }
            }
        }
    }
    return count($nesting) == 0;
}

File: go/matching-brackets/matching_brackets.go
// Matching Brackets example program
package matchingbrackets

import "slices"

// Test if a given input string has matching brackets (), [], and {}
// @pram input: The string to check
// @returns: True if matching, False if not.
func Bracket(input string) bool {
	var bracketStack []rune = []rune{}
	var pair map[rune]rune = map[rune]rune{
		'}': '{',
		')': '(',
		']': '[',
	}

	for _, char := range input {
		switch char {
		case '{', '(', '[':
			bracketStack = slices.Insert(bracketStack, 0, char)
		case '}', ')', ']':
			var match rune = pair[char]
			if len(bracketStack) > 0 && bracketStack[0] == match {
				bracketStack = bracketStack[1:] // pop
			} else {
				return false
			}
		}
	}

	return len(bracketStack) == 0
}
